# Bin Packing Lab

- Authors:
  - Akanksha Nehete, nehetea@mcmaster.ca
  - Anna Yang, yanga49@mcmaster.ca
  - Hamna Malik, malikh32@mcmaster.ca
- Group ID on Avenue: 19 
- Gitlab URL: https://gitlab.cas.mcmaster.ca/nehetea/l2-bin-packing

# Library Demonstration

In [112]:
from macpacking.reader import DatasetReader, BinppReader, JburkardtReader
from macpacking.model  import Online, Offline
import macpacking.algorithms.offline as offline
from macpacking.__init__ import WeightSet, WeightStream # remove this when submitting

In [113]:
# demonstrating the reader for binpp and binpp-hard
dataset = '_datasets/binpp/N1C1W1/N1C1W1_A.BPP.txt'
reader: DatasetReader = BinppReader(dataset)
print(f'Dataset: {dataset}')
print(f'  - Bin Capacity: {reader.offline()[0]}')
print(f'  - Objects to pack: {sorted(reader.offline()[1])}')

Dataset: _datasets/binpp/N1C1W1/N1C1W1_A.BPP.txt
  - Bin Capacity: 100
  - Objects to pack: [3, 7, 7, 10, 11, 13, 14, 17, 20, 21, 22, 23, 24, 25, 27, 28, 28, 29, 30, 30, 33, 33, 40, 40, 42, 44, 46, 49, 51, 52, 56, 61, 62, 67, 67, 69, 72, 74, 76, 85, 86, 87, 88, 91, 92, 92, 96, 96, 99, 99]


## T1: Implementation of Jburkardt Reader and Most Terrible Binpacking Algorithm
See Report Body under code for explanation

In [114]:
# implementation of reader for jburkardt
datasetc = '_datasets/jburkardt/p04_c.txt'
datasetw = '_datasets/jburkardt/p04_w.txt'
reader: DatasetReader = JburkardtReader(datasetw, datasetc)
print(f'Dataset: {datasetc}, {datasetw}')
print(f'  - Bin Capacity: {reader.offline()[0]}')
print(f'  - Objects to pack: {sorted(reader.offline()[1])}')

Dataset: _datasets/jburkardt/p04_c.txt, _datasets/jburkardt/p04_w.txt
  - Bin Capacity: 524
  - Objects to pack: [9, 9, 10, 10, 10, 10, 10, 10, 12, 12, 12, 37, 37, 46, 84, 85, 106, 106, 106, 106, 127, 127, 127, 127, 127, 252, 252, 252, 252, 252, 252, 252, 442]


In [115]:
# demonstrating online implementation of most terrible binpacking algorithm
import macpacking.algorithms.online as online
strategy: Online = online.MostTerrible()
result = strategy(reader.offline())
print(f'nb_bins = {len(result["solution"])}')
print(f'{sorted(result["solution"])}')

nb_bins = 33
[[9], [9], [10], [10], [10], [10], [10], [10], [12], [12], [12], [37], [37], [46], [84], [85], [106], [106], [106], [106], [127], [127], [127], [127], [127], [252], [252], [252], [252], [252], [252], [252], [442]]


## T2: Implementation and BenchMarking of Algorithms
See Report body for Explanation

In [116]:
# finding the optimal solution using baseline
import macpacking.algorithms.baseline as baseline
strategy: Offline = baseline.BenMaier()
result = strategy(reader.offline())
print(f'nb_bins = {len(result)}')
print(f'{sorted(result)}')

nb_bins = 8
[[10, 9, 9], [106, 106, 106, 85, 84, 37], [127, 127, 127, 106, 37], [252, 127, 127, 12], [252, 252, 10, 10], [252, 252, 12], [252, 252, 12], [442, 46, 10, 10, 10]]


In [117]:
# demonstrating usage of offline version of NextFit algorithm
import macpacking.algorithms.online as online
strategy: Offline = offline.NextFitDecreasing()
result = strategy(reader.offline())
print(f'nb_bins = {len(result["solution"])}')
print(f'{sorted(result["solution"])}')

nb_bins = 8
[[46, 37, 37, 12, 12, 12, 10, 10, 10, 10, 10, 10, 9, 9], [106, 106, 106, 85, 84], [127, 127, 127, 106], [252, 127, 127], [252, 252], [252, 252], [252, 252], [442]]


In [118]:
# demonstrating usage of online version of NextFit algorithm
import macpacking.algorithms.online as online
strategy: Online = online.NextFit()
result = strategy(reader.offline())
print(f'nb_bins = {len(result["solution"])}')
print(f'{sorted(result["solution"])}')

nb_bins = 8
[[10, 252, 127, 106, 10], [46, 12, 127, 10, 12, 252, 10], [85, 37, 252, 10], [106, 84, 252], [127, 37, 106, 127, 106], [252, 10, 9, 127, 12], [252, 252], [442, 9]]


In [119]:
# demonstrating usage of offline version of FirstFit algorithm
import macpacking.algorithms.offline as offline
strategy: Offline = offline.FirstFitDecreasing()
result = strategy(reader.online())
print(f'nb_bins = {len(result["solution"])}')
print(f'{sorted(result["solution"])}')

nb_bins = 7
[[106, 106, 106, 85, 84, 37], [127, 127, 127, 106, 37], [252, 127, 127, 9, 9], [252, 252, 10, 10], [252, 252, 10, 10], [252, 252, 10, 10], [442, 46, 12, 12, 12]]


In [120]:
# demonstrating usage of online version of FirstFit algorithm
import macpacking.algorithms.online as online
strategy: Online = online.FirstFit()
result = strategy(reader.online())
print(f'nb_bins = {len(result["solution"])}')
print(f'{sorted(result["solution"])}')

nb_bins = 8
[[10, 252, 127, 106, 10, 12], [46, 127, 12, 252, 9, 37, 10, 10, 9, 12], [85, 252, 127], [106, 84, 252], [127, 37, 106, 127, 106, 10, 10], [252], [252, 252], [442]]


In [121]:
# demonstrating usage of offline version of BestFit algorithm
import macpacking.algorithms.offline as offline
strategy: Offline = offline.BestFitDecreasing()
result = strategy(reader.online())
print(f'nb_bins = {len(result["solution"])}')
print(f'{sorted(result["solution"])}')

nb_bins = 8
[[10, 9, 9], [106, 106, 106, 85, 84, 37], [127, 127, 127, 106, 37], [252, 127, 127, 12], [252, 252, 10, 10], [252, 252, 12], [252, 252, 12], [442, 46, 10, 10, 10]]


In [122]:
# demonstrating usage of online version of BestFit algorithm
import macpacking.algorithms.online as online
strategy: Online = online.BestFit()
result = strategy(reader.online())
print(f'nb_bins = {len(result["solution"])}')
print(f'{sorted(result["solution"])}')

nb_bins = 8
[[10, 252, 127, 106, 10, 12], [46, 127, 12, 252, 84], [106, 252, 85], [127, 37, 106, 127, 106, 10, 10], [127, 252], [252], [252, 252, 10, 9], [442, 9, 37, 10, 12]]


In [123]:
# demonstrating usage of offline version of WorstFit algorithm
import macpacking.algorithms.offline as offline
strategy: Offline = offline.WorstFitDecreasing()
result = strategy(reader.online())
print(f'nb_bins = {len(result["solution"])}')
print(f'{sorted(result["solution"])}')

nb_bins = 8
[[9, 9], [106, 106, 106, 85, 84, 37], [127, 127, 127, 106, 37], [252, 127, 127, 10], [252, 252, 10, 10], [252, 252, 10, 10], [252, 252, 12], [442, 46, 12, 12, 10]]


In [124]:
# demonstrating usage of online version of WorstFit algorithm
import macpacking.algorithms.online as online
strategy: Online = online.WorstFit()
result = strategy(reader.online())
print(f'nb_bins = {len(result["solution"])}')
print(f'{sorted(result["solution"])}')

nb_bins = 8
[[10, 252, 127, 106, 10], [46, 12, 127, 10, 12, 252, 10], [85, 37, 252, 10, 12], [106, 84, 252], [127, 37, 106, 127, 106], [252, 10, 9, 127], [252, 252], [442, 9]]


## T3: Measuring Improvement Margin
See Report body for further explanation of graphs etc. The graph will open in a new window, so thats why it's not showing in the Jupyter notebook

In [125]:
from improvement_margin import Margin
from macpacking.algorithms.online import NextFit, MostTerrible, WorstFit, BestFit, RefinedFirstFit
from macpacking.algorithms.offline import NextFitDecreasing, WorstFitDecreasing, BestFitDecreasing, RefinedFirstFitDecreasing

# Margin class for the determination of of improvement margins for a specified algorithm 
test_binpp = Margin('optimal_solutions/binpp.csv', NextFit(), 'Next Fit', True, 'N1C1W1')
# displays continous results, i.e graph representing the difference of bins from optimal (the graph is displayed in a new window)
# ** CLOSE THE GRAPH WINDOW SO THE CELL CAN CONTINUE RUNNING
#test_binpp.display_continuous()
# displays discrete results, i.e whether the solution is optimal or not
test_binpp.display_discrete()



N1C1W1_A is not optimal
N1C1W1_B is not optimal
N1C1W1_C is not optimal
N1C1W1_D is not optimal
N1C1W1_E is not optimal
N1C1W1_F is not optimal
N1C1W1_G is not optimal
N1C1W1_H is not optimal
N1C1W1_I is not optimal
N1C1W1_J is not optimal
N1C1W1_K is not optimal
N1C1W1_L is not optimal
N1C1W1_M is not optimal
N1C1W1_N is not optimal
N1C1W1_O is not optimal
N1C1W1_P is not optimal
N1C1W1_Q is not optimal
N1C1W1_R is not optimal
N1C1W1_S is not optimal
N1C1W1_T is not optimal


## T4: Add Smarter Algorithms
See Report body for further explation of algortithm and KPI graphs.

In [126]:
# demonstrating usage of online version of the Refined First Fit algorithm
import macpacking.algorithms.online as online
strategy: Online = online.RefinedFirstFit()
result = strategy(reader.online())
print(f'nb_bins = {len(result["solution"])}')
print(f'{sorted(result["solution"])}')

nb_bins = 8
[[10, 127, 106, 10, 127, 37, 106], [106, 84, 85, 9, 127, 12], [127, 106, 46, 12, 127, 10, 12, 10, 9, 37, 10, 10], [252], [252, 252], [252, 252], [252, 252], [442]]


In [127]:
# demonstrating usage of offline version of the Refined First Fit algorithm
import macpacking.algorithms.offline as offline
strategy: Offline = offline.RefinedFirstFitDecreasing()
result = strategy(reader.online())
print(f'nb_bins = {len(result["solution"])}')
print(f'{sorted(result["solution"])}')

nb_bins = 8
[[106, 85, 84, 37, 37, 10, 10, 10, 10, 10, 10, 9], [127, 106, 106, 106, 46, 12, 12, 9], [127, 127, 127, 127, 12], [252], [252, 252], [252, 252], [252, 252], [442]]


## T5: From Fixed Capacity to Fixed Bins
See Report body for further explation of algorithm and KPI graphs.

In [128]:
# demonstrating the baseline integrated using the to_constant_volume() function in the binpacking 
# library that attempts to solve the problem of multiway number partitioning
import macpacking.algorithms.baseline as baseline
# initializing number of bins by passing argument into ConstantBinNumber
strategy: Offline = baseline.ConstantBinNumber(num_bins=8)
result = strategy(reader.offline())
print(f'nb_bins = {len(result)}')
print(f'{sorted(result)}')

nb_bins = 8
[[252, 106, 106], [252, 106, 106], [252, 127, 37, 12, 10, 10], [252, 127, 37, 12, 10, 10, 9], [252, 127, 46, 12, 10, 9], [252, 127, 84], [252, 127, 85], [442, 10]]


In [129]:
# demonstrating the Greedy Algorithm compares with the baseline for multiway number partitioning
import macpacking.algorithms.online as online
strategy: Offline = offline.GreedyHeuristic(num_bins=8)
result = strategy(reader.offline())
print(f'nb_bins = {len(result["solution"])}')
print(f'{sorted(result["solution"])}')

nb_bins = 8
[[10, 106, 10, 9, 252], [10, 127, 85, 252], [37, 46, 12, 127, 10, 252], [106, 10, 252], [106, 12, 442], [127, 84, 37, 127], [127, 106, 10, 9, 252], [252, 12, 252]]


# Report

### T1 Explanation 

#### Dataset Explanation
There are three datasets included in the repository: binpp, binpp-hard, and jburkardt. Binpp and binpp-hard have the same format, they are given as a long text file in which every line is an integer. The first number represents the number of objects that need to be packed, and the second line represents the maximum weight or capacity a bin can hold. Every subsequent line in the file represents the weight of an object that needs to be packed. Let n be the number of objects that need to be packed and let c be the maximum capacity of the bin. In binpp, n = 50, 100, 200, 500 and c = 100, 120, and 150. Binpp-hard has the same formatting, but the values of n = 200 and c = 100000. In the jburkhardt dataset, the file system is arranged a little differently than the binpp one as there are 3 files for every problem. For problems p01, p02, and p03, c = 100 and n ranges from 9 to 14. For problem p04, c = 524 and n = 33. For each problem, the file ending with c contains the capacity number and the file ending with w contains the weights of the objects that need to be packed. The s file contains an assignment of weights. The values of the n parameter are important to evaluate algorithm performance, as the time complexity of the problem is in terms of n. The longer n is, the longer the algorithm will take to find the number of bins needed to optimally pack the n items.

#### SOLID Principles Explanation of UML Diagram

- Single responsibility
The separation of online and offline algorithms implements the single responsibility principle, as all Online and Offline algorithms will be separated into their respective files.
Each implementation of a bin packing algorithm is given its own class with its own functions it will use to perform bin packing. By separating each algorithm into its own class with its own functions, the single responsibility principle is followed.
- Open closed
The use of a BinPacker interface for the bin packing algorithms as well as the separation of Online and Offline algorithms uses the open-closed principle. The program is open to extension of new algorithm classes, but it is closed to modification as BinPacker, Online, and Offline need not be modified in order to accommodate for the addition of algorithms.
- Interface segregation
Interface segregation was used by the implementation of the BinPacker interface. This interface separates Online and Offline bin packing algorithms, allowing the user to access any of the Online or Offline implementations easily.
- Dependency injection
The details of each bin packing algorithm implementation are unimportant to the user, and thus, the user does not need to be made aware of the internal functions that are ‘injected’ in each algorithm class. Only the _process() function of each algorithm matters, which is why this design implants dependency injection.

#### Algorithms Explanation and Complexity Analysis

- Next Fit
This is the simplest approximate approach to the bin packing problem. This is an online algorithm in which the first item is assigned to the first bin. The weight stream is then considered by increasing indexes. If an item fits in the current bin, it is assigned to that bin. Otherwise, the next bin becomes the current one. The time complexity of the algorithm is O(n), as the algorithm only traverses the weight stream once.

- Next Fit Decreasing
This is an offline algorithm in which the weights are sorted first into decreasing order. Then, the Next Fit algorithm is applied as discussed above. The time complexity of the algorithm would be O(n + nlogn) which approximates to O(nlogn) since it will take O(nlogn) time to sort the list of weights first using timsort (Python’s default sorting algorithm).

- First Fit 
This is an online algorithm which considers the weight stream according to increasing indices. It assigns each item to the lowest indexed initialized bin into which it fits. When the current item cannot fit into any of the previous initialized bins, a new bin is introduced and the item is assigned to that bin. The time complexity of the algorithm is O(n2), because it has to traverse the weight stream, but also has to loop through previous bins if needed. However, this time complexity can be reduced to O(nlogn) if a self balancing binary tree is used as the data structure.

- First Fit Decreasing
This is an offline algorithm in which the weights are sorted first into decreasing order. Then, the First Fit algorithm is applied as discussed above. The time complexity of the algorithm would be O(n2 + nlogn) which approximates to O(n2) since it will take O(nlogn) time to sort the list of weights first using timsort (Python’s default sorting algorithm). 

- Best Fit
This is an online algorithm which considers the weight stream according to increasing indices. It  assigns each item to the feasible bin having the smallest space left but can still fit the item. Essentially, it will place each item in the tightest spot. The time complexity of the algorithm is O(n2), because it has to traverse the weight stream, but also has to loop through previous bins if needed in order to determine the tightest spot.

-Best Fit Decreasing
This is an offline algorithm in which the weights are sorted first into decreasing order. Then, the Best Fit algorithm is applied as discussed above. The time complexity of the algorithm would be O(n2 + nlogn) which approximates to O(n2) since it will take O(nlogn) time to sort the list of weights first using timsort (Python’s default sorting algorithm).

- Worst Fit
This is an online algorithm which considers the weight stream according to increasing indices. It assigns each item to the feasible bin having the most empty space. Essentially, it will place the item in the least tightest spot to even out the bins. The time complexity of the algorithm is O(n2), because it has to traverse the weight stream, but also has to loop through previous bins if needed in order to determine the least tight spot.

- Worst Fit Decreasing
This is an offline algorithm in which the weights are sorted first into decreasing order. Then, the Worst Fit algorithm is applied as discussed above. The time complexity of the algorithm would be O(n2 + nlogn) which approximates to O(n2) since it will take O(nlogn) time to sort the list of weights first using timsort (Python’s default sorting algorithm).

#### T2: KPI Explanation and Analysis

![title]("graphs/comparisons_all.png")

## Self-reflection questions

As part of the self-reflection dimension of an experiential course, each member of the group is expected to answer to the following four questions:

  - What process did you go through to produce this result? (Backward)
  - What were your standards for this piece of work? Did you meet your standards? (Inward)
  - What the one thing you particularly want people to notice when they look at your work? (Outward)
  - What lessons will you keep from this reading/lecture in your professional practice? (Forward)